storage/innobase/xtrabackup/src/xbcloud/xbcloud.h

```diff
@@ -0,0 +1,186 @@
+/******************************************************
+Copyright (c) 2014, 2022 Percona LLC and/or its affiliates.
+
+xbcloud utility. Manage backups on cloud storage services.
+
+This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; version 2 of the License.
+
+This program is distributed in the hope that it will be useful,
+but WITHOUT ANY WARRANTY; without even the implied warranty of
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program; if not, write to the Free Software
+Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
+
+*******************************************************/
+
+struct file_metadata_t {
+  my_off_t last_chunk;
+  my_off_t next_chunk;
+  std::string name;
+};
+
+/** struct to hold state of current in progress files. This avoids multiple
+ * chunks from the same file to be downloaded in parallel */
+struct download_state {
+  std::mutex m;
+  /* files with active chunks been downloaded (waiting http client to complete
```

Comment:
can there be multiple threads downloading multiple chunks in same file?

Reply:
Not possible, that is ensured by download_state, we get a file from global list of files and that file remain bind to the thread that first assigned it.